File: cli/src/lib/projectPreview.ts
import { execFile } from 'node:child_process'
import { open, opendir, realpath, stat } from 'node:fs/promises'
import { constants } from 'node:fs'
import { homedir } from 'node:os'
import { isAbsolute, join, relative } from 'node:path'
import { promisify } from 'node:util'
import { env } from '../config/env.js'

const exec = promisify(execFile)
const within = (root: string, path: string) => {
  const rel = relative(root, path)
  return rel === '' || (rel !== '..' && !rel.startsWith('../') && !isAbsolute(rel))
}

/** Existing local content only. Paths/README symlinks stay inside the browsable
 * home or an agent's known workspace; all Git commands are read-only/bounded. */
export async function projectPreview(path: string, knownRoots: string[] = []) {
  if (!isAbsolute(path) || path.length > 4096 || /[\x00-\x1f\x7f]/.test(path)) return { error: 'INVALID_PATH' }
  try {
    const target = await realpath(path)
    const roots = await Promise.all([homedir(), ...knownRoots].map(root => realpath(root).catch(() => '')))
    if (env.HARNESS_FS_BROWSE_UNRESTRICTED !== '1' && !roots.some(root => root && within(root, target))) return { error: 'FORBIDDEN' }
    if (!(await stat(target)).isDirectory()) return { error: 'NOT_FOUND' }
    const files: string[] = []
    let readme: string | undefined
    const directory = await opendir(target)
    let seen = 0
    for await (const entry of directory) {
      if (!entry.name.startsWith('.')) files.push(entry.name)
      if (!readme && entry.isFile() && /^readme(?:\.(?:md|markdown|txt))?$/i.test(entry.name)) {
        const file = await open(join(target, entry.name), constants.O_RDONLY | constants.O_NOFOLLOW)
        try {
          const buffer = Buffer.alloc(32 * 1024)
          const { bytesRead } = await file.read(buffer, 0, buffer.length, 0)
          const bytes = buffer.subarray(0, bytesRead)
          if (!bytes.includes(0)) readme = bytes.toString('utf8')
        } finally { await file.close() }
      }
      if (++seen >= 200) break
    }
    const git = async (args: string[]) => {
      try {
        return (await exec('git', ['--no-optional-locks', '-c', 'core.fsmonitor=false', '-C', target, ...args], {
          timeout: 2000, maxBuffer: 32 * 1024,
          env: { ...process.env, GIT_TERMINAL_PROMPT: '0', GIT_OPTIONAL_LOCKS: '0' },
        })).stdout.trim()
      } catch { return undefined }
    }
    const [branch, status, last, authors] = await Promise.all([
      git(['symbolic-ref', '--quiet', '--short', 'HEAD']),
      git(['status', '--porcelain=v1', '--untracked-files=no']),
      git(['log', '-1', '--format=%s%n%an%n%aI']),
      git(['log', '-100', '--format=%an']),
    ])
    const commit = last?.split('\n')
    return { path: target, readme, files: files.sort().slice(0, 12), branch,
      ...(status !== undefined ? { changedFiles: status ? status.split('\n').length : 0 } : {}),
      ...(commit && commit.length >= 3 ? { commit: { subject: commit[0], author: commit[1], date: commit[2] } } : {}),
      contributors: authors ? [...new Set(authors.split('\n'))].slice(0, 5) : [],
    }
  } catch { return { error: 'UNREADABLE' } }
}

File: desktop/lib/core/project_preview.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';

import 'package:path/path.dart' as p;

/// Read-only, bounded source material for a project preview. Never executes
/// project code, hooks, Markdown images or a Git network operation.
Future<Map<String, dynamic>> readLocalProjectPreview(String path) async {
  if (!p.isAbsolute(path) || path.length > 4096 || path.contains('\u0000')) {
    return {'error': 'INVALID_PATH'};
  }
  try {
    if (!await Directory(path).exists()) return {'error': 'NOT_FOUND'};
    final files = <String>[];
    String? readme;
    var seen = 0;
    await for (final entity in Directory(path).list(followLinks: false)) {
      final name = p.basename(entity.path);
      if (!name.startsWith('.')) files.add(name);
      if (readme == null &&
          RegExp(
            r'^readme(?:\.(?:md|markdown|txt))?$',
            caseSensitive: false,
          ).hasMatch(name) &&
          entity is File) {
        final file = await entity.open();
        try {
          final bytes = await file.read(32 * 1024);
          if (!bytes.contains(0)) {
            readme = utf8.decode(bytes, allowMalformed: true);
          }
        } finally {
          await file.close();
        }
      }
      if (++seen >= 200) break;
    }
    files.sort();
    final git = await Future.wait([
      _git(path, ['symbolic-ref', '--quiet', '--short', 'HEAD']),
      _git(path, ['status', '--porcelain=v1', '--untracked-files=no']),
      _git(path, ['log', '-1', '--format=%s%n%an%n%aI']),
      _git(path, ['log', '-100', '--format=%an']),
    ]);
    final commit = git[2]?.split('\n');
    return {
      'path': path,
      'readme': ?readme,
      'files': files.take(12).toList(),
      'branch': ?git[0],
      if (git[1] != null)
        'changedFiles': git[1]!.isEmpty ? 0 : git[1]!.split('\n').length,
      if (commit != null && commit.length >= 3)
        'commit': {
          'subject': commit[0],
          'author': commit[1],
          'date': commit[2],
        },
      if (git[3] != null)
        'contributors': git[3]!
            .split('\n')
            .where((name) => name.isNotEmpty)
            .toSet()
            .take(5)
            .toList(),
    };
  } on FileSystemException {
    return {'error': 'UNREADABLE'};
  }
}

Future<String?> _git(String path, List<String> args) async {
  Process? process;
  Timer? timer;
  var expired = false;
  try {
    process = await Process.start(
      'git',
      [
        '--no-optional-locks',
        '-c',
        'core.fsmonitor=false',
        '-C',
        path,
        ...args,
      ],
      environment: {'GIT_TERMINAL_PROMPT': '0', 'GIT_OPTIONAL_LOCKS': '0'},
    );
    timer = Timer(const Duration(seconds: 2), () {
      expired = true;
      process?.kill(ProcessSignal.sigkill);
    });
    var length = 0;
    final bytes = <int>[];
    final stdout = process.stdout.listen((chunk) {
      length += chunk.length;
      if (length > 32 * 1024) {
        expired = true;
        process?.kill(ProcessSignal.sigkill);
      } else {
        bytes.addAll(chunk);
      }
    });
    final outputDone = stdout.asFuture<void>();
    final stderr = process.stderr.drain<void>();
    final code = await process.exitCode;
    await outputDone;
    await stderr;
    return code == 0 && !expired
        ? utf8.decode(bytes, allowMalformed: true).trim()
        : null;
  } on ProcessException {
    return null;
  } finally {
    timer?.cancel();
  }
}
